### ISSUE TITLE
Aggregate function `List()` (`Array_Agg()`) does not work with `Unnest()`

### ISSUE THREAD

**Issue by manticore-projects** [duckdb/duckdb#11457]
### What happens?

When applying `List()` on a `Unnest()` it will return strange `Struct`? instead of an Array of elements. The result can not be accessed or cast to an Array either.

### To Reproduce

```sql
-- technically works but produces wrong output
SELECT list( x )[1] element FROM UNNEST([2, 1,-2, 3, -2, 1, 2]) AS x;
```
```
/*
┌────────────────────────────────────────────┐
│ element                                    │
├────────────────────────────────────────────┤
│ {main.list_value(2, 1, -2, 3, -2, 1, 2)=2} │
└────────────────────────────────────────────┘
*/
```
```sql
-- fails because the result of `List()` is not accessible
SELECT list( x )[1] element, list( x )[1] = 2 test1 FROM UNNEST([2, 1,-2, 3, -2, 1, 2]) AS x;
```
```sql
-- This is what SHOULD happen (because its similar):
SELECT list( x )[1] element, list( x )[1] = 2 test1 FROM (SELECT 2 x UNION ALL SELECT 1);
```
```
/*
┌─────────┬───────┐
│ element │ test1 │
├─────────┼───────┤
│       2 │ true  │
└─────────┴───────┘
*/
```

### OS:

Linux

### DuckDB Version:

0.11-Snapshot

### DuckDB Client:

Java

### Full Name:

Andreas Reichel

### Affiliation:

manticore-projects.com

### Have you tried this on the latest [nightly build](https://duckdb.org/docs/installation/?version=main)?

I have tested with a nightly build

### Have you tried the steps to reproduce? Do they include all relevant data and configuration? Does the issue you report still appear there?

- [X] Yes, I have

**Comment by soerenwolfers** [comment]
Same as in [this other one ticket or yours](https://github.com/duckdb/duckdb/issues/11335) , you're confusing naming a column and naming a dataframe.

**Comment by manticore-projects** [comment]
> Same as in [this other one ticket of yours](https://github.com/duckdb/duckdb/issues/11335) , you're confusing naming a column and naming a dataframe.

Thank you @soerenwolfers, only now with this particular phrasing I actually do understand the problem. I am grateful for this enlightenment.

However, I still believe that those reports have merit because:

1) there is no documentation  on the use of `Unnest` as a data frame, the page explains only `Unnest` as a column
2) you yourself refer to the word "magic"
3) It "just works" for many other RDBMS and this fools end-users easily into this debate especially when there is not documentation.

In a Nutshell, I understand now that:

``` sql
-- generates a Data Frame `x` without any useful column name
SELECT list( x )[1] element FROM UNNEST([2, 1,-2, 3, -2, 1, 2]) AS x;

-- generates a Data Frame `x` with a usable column name `y`
SELECT list( x )[1] element FROM (select UNNEST([2, 1,-2, 3, -2, 1, 2]) AS y) as x;
```

This should be explained somewhere and also the should be a more elegant way to assign a column name to the data frame. I don't think I will be the only user to be confused much about this.

But again, thanks for your explanation and patience!
How can I help with the documentation in the most effective way please?

**Comment by soerenwolfers** [comment]
The only thing that's magic here is that you can use the name of a dataframe where you'd usually use a column name, and it'll give you a column of struct type. 

I am sure I've seen this in the documentation somewhere, but can't remember where. 

Everything else, including applying a function like unnest, is the same as always.

Regarding a more elegant way to assign a column name: typically you'd name table and column at once, via `... FROM <table expression> table name(columnName1, ..., columnName n)`

**Comment by manticore-projects** [comment]
Again, thank you for your time and effort and also the patience.

> I am sure I've seen this in the documentation somewhere, but can't remember where.

The `Unnest()` related page does not say anything about a dataframe, but only explains columns.

At the same time, there is https://duckdb.org/docs/sql/query_syntax/from#syntax which explains "TableFunction" as "JOIN TABLEFUNCTION() AS T(x,y)" which is indeed common syntax.

Although it won't work on UNNEST():

```sql

select t, len from Unnest(['abc', 2]) AS T(t, len)
```

In my understanding, a direct contradiction of the documentation.
And no better clarification anywhere to find (except your kind explanations in this issue).

> Regarding a more elegant way to assign a column name: typically you'd name table and column at once, via `... FROM <table expression> tableName(columnName1, ..., columnName n)`

Yes, and how would that look correctly on `Unnest()' please?

> Finally, do you have an example where another RDBMS "just works" and duckdb doesn't?

Sure:  https://cloud.google.com/bigquery/docs/reference/standard-sql/aggregate_functions#count

Check the examples, which all will work on DuckDB also -- as long as you rewrite the Unnest() function.

**Comment by manticore-projects** [comment]
Good Morning.

> In your `--expected duckdb` example why are you naming both the column and the table (and even worse name them both "x")? Just name the column and it doesn't look so bad anymore, does it?

I agree of course, that in this illustration I better would have used:

```sql
-- expected DuckDB
SELECT COUNT(DISTINCT IF(x > 0, x, NULL)) AS distinct_positive
FROM (SELECT UNNEST([1, -2, 4, 1, -5, 4, 1, 3, -6, 1]) AS COLUMN_ALIAS) AS TABLE_ALIAS;
```
And I would like to recommend to use an example like this for the documentation (unless I am indeed the only one having a problem here.)


> Still, I am not sure whether the syntax of your BigQuery example or your `FROM unnest(...) as T(t, len)` should work in duckdb, but I think I agree with you that at least one of them should work.

Thank you, this is exactly my main problem. When I read:
``` sql
select ... from UNNEST(...) ALIAS;
```

then I do understand that this is a Table Alias (on a data frame). So far so good.
However, this data frame will have exactly 1 column only (unless recursive is used).

And so indeed it would make sense to:
1) access this column,  when the Table Alias is provided -- or at least throw an error, when the Data Frame is provided instead of the Column Reference.
2) allow Table Aliases with Column Names `T(x)`
3) explain and document all of this somewhere ( I would volunteer on this of course)

**Comment by soerenwolfers** [comment]
Neither of your suggestions in 1.  are easily squarable agree with existing duckdb logic:  table names are already valid column identifiers (and provides a column of struct type). I don't see why that should be changed.

**Comment by cmdlineluser** [comment]
**Disclaimer:** I'm a novice, so this is very much open for correction.

From what I can tell, the usage of  SELECTing the "table_name" resulting in a STRUCT appears to be explained by [Composite Types in Postgres.](https://www.postgresql.org/docs/current/rowtypes.html)

> Whenever you create a table, a composite type is also automatically created, with the same name as the table, to represent the table's row type.

```sql
-- POSTGRES
CREATE TABLE tbl (A int, B int, C int);
INSERT INTO tbl VALUES (1, 2, 3), (4, 5, 6);
SELECT pg_typeof(tbl), tbl FROM tbl;
```

```sql
 pg_typeof |   tbl   
-----------+---------
 tbl       | (1,2,3)  -- STRUCT in DuckDB {'a': 1, 'b': 2, 'c': 3}
 tbl       | (4,5,6)  -- STRUCT in DuckDB {'a': 4, 'b': 5, 'c': 6}
``` 

It looks like this is where the STRUCTs come from, but I can't seem to find anything in the docs.

**Comment by cmdlineluser** [comment]
@soerenwolfers I think I may have stumbled upon the doc entry you were referring to:

- https://duckdb.org/docs/sql/expressions/subqueries#returning-each-row-of-the-subquery-as-a-struct

**Comment by soerenwolfers** [comment]
Yes, the very bottom of that page was what I was referring to (the page structure is a bit unclear that this is not specific to correlated subqueries)
